### PR TITLE
Promo tweet WIP

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: app/assets/javascripts/components/dashboard/navigation/content_panel/content_panel_views/promote/file_store_ref.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
+var FileStore = (function($) {
```
Comment (signalplus-dev) [inline app/assets/javascripts/components/dashboard/navigation/content_panel/content_panel_views/promote/file_store_ref.js]: Use `const`.

Comment (signalplus-dev) [inline app/assets/javascripts/components/dashboard/navigation/content_panel/content_panel_views/promote/file_store_ref.js]:
``` js
export default const FileStore
```

File: app/assets/javascripts/components/dashboard/navigation/content_panel/content_panel_views/promote/promo_tweet_image.jsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  componentDidMount() {
+    FileStore.getResources()
+    .then(function(data) {
+      this.setState({ images: data.images });
+    }.bind(this));
+  }
+
+  handleCreateDocument(document) {
+    this.setState({ documents: $.merge([document], this.state.documents) });
+  },
```
Comment (signalplus-dev) [inline app/assets/javascripts/components/dashboard/navigation/content_panel/content_panel_views/promote/promo_tweet_image.jsx]: No comma.

File: app/services/promotional_tweet/create_promotional_tweet.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  attr_reader :current_user, :params, :promotional_tweet
+
+  # @param params [Hash]
+  def initialize(params)
+    @params = params
+    build
+  end
+
+  # @return [Boolean] PromotionalTweet saved
+  def call
```
Comment (signalplus-dev) [inline app/services/promotional_tweet/create_promotional_tweet.rb]: Is `call` a reserved method?

File: Gemfile [hunk 4]
```diff
@@ -29,7 +29,8 @@ gem 'sidetiq'
 gem 'sidekiq-unique-jobs'
 gem 'sidekiq-failures'
 gem 'sinatra'
-gem 'aws-sdk', '~> 2'
+gem 'aws-sdk-v1'
```
Comment (signalplus-dev) [inline Gemfile]: Do we need this?

Comment (ghost) [reply Gemfile]: Ah I had some trouble trying to get pre-signed url to pass to component with V2, I'll mess around a bit more and use v2

File: .gitignore [hunk 5]
```diff
@@ -14,3 +14,4 @@
 __init__.py
 application.yml
 node_modules/
+.rdb
```
Comment (signalplus-dev) [inline .gitignore]: 👏

File: app/controllers/api/v1/promotional_tweets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    @promotional_tweet = @create_promotional_tweet_service.promotional_tweet
+    if @create_promotional_tweet_service.call
+      render :show, status: :created
+    else
+      render json: { errors: @promotional_tweet.errors.messages }, status: :unprocessable_entity
+    end
+  end
+
+  def post_tweet
+    binding.pry
```
Comment (signalplus-dev) [inline app/controllers/api/v1/promotional_tweets_controller.rb]: Remove this.

File: app/controllers/api/v1/uploads_controller.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+class Api::V1::UploadsController < ApplicationController
```
Comment (signalplus-dev) [inline app/controllers/api/v1/uploads_controller.rb:1]: This should eventually be changed to inherit from the `Api::V1::BaseController`. That'll handle the authentication. That can be done when you change the front end to use the new fetch implementation.

File: app/models/promotional_tweet.rb [hunk 8]
```diff
@@ -16,10 +16,24 @@
...
+  DIRECT_UPLOAD_URL_FORMAT = /\A(https:\/\/signal-twitter-images-development.s3.amazonaws.com\/promotional-images\/.+\/.+\.(jpg|png|jpeg))\z/
+
+  enum status: { unprocessed: 0, processed: 1 }
+  has_attached_file :image
+  validates_format_of :direct_upload_url, with: DIRECT_UPLOAD_URL_FORMAT
+  do_not_validate_attachment_file_type :image
+
+
+  def self.update_or_create_by(args, attributes)
+    binding.pry
```
Comment (signalplus-dev) [inline app/models/promotional_tweet.rb]: Remove this.

File: app/services/promotional_tweet/generate_upload_url.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  #
+  # @param filename [String]
+  #
+  def initialize(image_filename)
+    @image_filename = image_filename
+    @content_type   = MIME::Types.type_for(image_filename).first.content_type
+  end
+
+  # @return [Boolean] Generation successful
+  def call
```
Comment (signalplus-dev) [inline app/services/promotional_tweet/generate_upload_url.rb]: Let's not use `#call` as a method as it is a base method in Ruby.

File: app/workers/kiqs/process_promotional_tweet_worker.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+class ProcessPromotionalTweetWorker
+  include Sidekiq::Worker
+
+  def perform(promotional_tweet)
```
Comment (signalplus-dev) [inline app/workers/kiqs/process_promotional_tweet_worker.rb]: You can't pass full objects in `Sidekiq` workers. You can pass it as YAML but that's not highly recommended. Can we pass it in as a hash of attributes to create the promotional_tweet?

**Comment by ghost** [comment]
@ricardo-quinones

**Comment by ghost** [comment]
@project-signal yooo rick am i good to go now?

**Comment by signalplus-dev** [comment]
👍

**Pull request merged**